File: clients/Cpp/model/DebugState.cpp
#include "DebugState.hpp"

DebugState::DebugState() { }
DebugState::DebugState(Vec2Int windowSize, Vec2Float mousePosWindow, Vec2Float mousePosWorld, std::vector<std::string> pressedKeys, Camera camera, int playerIndex) : windowSize(windowSize), mousePosWindow(mousePosWindow), mousePosWorld(mousePosWorld), pressedKeys(pressedKeys), camera(camera), playerIndex(playerIndex) { }
DebugState DebugState::readFrom(InputStream& stream) {
    DebugState result;
    result.windowSize = Vec2Int::readFrom(stream);
    result.mousePosWindow = Vec2Float::readFrom(stream);
    result.mousePosWorld = Vec2Float::readFrom(stream);
    result.pressedKeys = std::vector<std::string>(stream.readInt());
    for (size_t i = 0; i < result.pressedKeys.size(); i++) {
        result.pressedKeys[i] = stream.readString();
    }
    result.camera = Camera::readFrom(stream);
    result.playerIndex = stream.readInt();
    return result;
}
void DebugState::writeTo(OutputStream& stream) const {
    windowSize.writeTo(stream);
    mousePosWindow.writeTo(stream);
    mousePosWorld.writeTo(stream);
    stream.write((int)(pressedKeys.size()));
    for (const std::string& pressedKeysElement : pressedKeys) {
        stream.write(pressedKeysElement);
    }
    camera.writeTo(stream);
    stream.write(playerIndex);
}

File: clients/Go/model/debug_state.go
package model

import "io"
import . "aicup2020/stream"

type DebugState struct {
    WindowSize Vec2Int32
    MousePosWindow Vec2Float32
    MousePosWorld Vec2Float32
    PressedKeys []string
    Camera Camera
    PlayerIndex int32
}
func NewDebugState(windowSize Vec2Int32, mousePosWindow Vec2Float32, mousePosWorld Vec2Float32, pressedKeys []string, camera Camera, playerIndex int32) DebugState {
    return DebugState {
        WindowSize: windowSize,
        MousePosWindow: mousePosWindow,
        MousePosWorld: mousePosWorld,
        PressedKeys: pressedKeys,
        Camera: camera,
        PlayerIndex: playerIndex,
    }
}
func ReadDebugState(reader io.Reader) DebugState {
    result := DebugState {}
    result.WindowSize = ReadVec2Int32(reader)
    result.MousePosWindow = ReadVec2Float32(reader)
    result.MousePosWorld = ReadVec2Float32(reader)
    result.PressedKeys = make([]string, ReadInt32(reader))
    for i := range result.PressedKeys {
        result.PressedKeys[i] = ReadString(reader)
    }
    result.Camera = ReadCamera(reader)
    result.PlayerIndex = ReadInt32(reader)
    return result
}
func (value DebugState) Write(writer io.Writer) {
    value.WindowSize.Write(writer)
    value.MousePosWindow.Write(writer)
    value.MousePosWorld.Write(writer)
    WriteInt32(writer, int32(len(value.PressedKeys)))
    for _, PressedKeysElement := range value.PressedKeys {
        WriteString(writer, PressedKeysElement)
    }
    value.Camera.Write(writer)
    WriteInt32(writer, value.PlayerIndex)
}
